src/parser/jobs/blm/Gauge.js

```diff
@@ -7,210 +7,291 @@ import ACTIONS from 'data/ACTIONS'
+	_hasEnochian = false
+	_enochianTimer = 0
+	_hasPolyglot = false
+
 	_droppedEno = 0
 	_lostFoul = 0
 	_overwrittenFoul = 0
-	_beginOfCast = 0
 
-	constructor(...args) {
-		super(...args)
-		this.addHook('begincast', {by: 'player'}, this._onBegin)
-		this.addHook('cast', {by: 'player'}, this._onCast)
-		this.addHook('death', {to: 'player'}, this._onDeath)
-		this.addHook('complete', this._onComplete)
-	}
+	_normalizeIndex = 0
+	_currentTimestamp = 0
 
-	enoTimerCheck(event) {
-		const AFUIRunTime = event.timestamp - this._AFUITimer
-
-		//reseting AF/UI and dropping eno due to going past the timer
-		if (AFUIRunTime > ASTRAL_UMBRAL_DURATION) {
-			if (this._eno) {
-				this._eno = false
-				this._enoTimer = 0
-				this._droppedEno ++
-				if (this._poly === 0) {
-					this._lostFoul ++
-				}
-			}
-			this._AF = 0
-			this._UI = 0
-			this._UH = 0
-			this._AFUITimer = 0
+	_toAdd = []
+	_lastAdded = null
+
+	gaugeValuesChanged(lastGaugeEvent) {
+		if (!lastGaugeEvent) {
+			return true
 		}
+		if (lastGaugeEvent.astralFire !== this._astralFireStacks ||
+			lastGaugeEvent.umbralIce !== this._umbralIceStacks ||
+			lastGaugeEvent.umbralHearts !== this._umbralHeartStacks ||
+			lastGaugeEvent.enochian !== this._hasEnochian ||
+			lastGaugeEvent.polyglot !== this._hasPolyglot
+		) {
+			return true
+		}
+		return false
 	}
 
-	enoDrop() {
-		this._UI = 0
-		this._AF = 0
-		this._UH = 0
-		if (this._poly === 0 && this._eno) {
-			this._eno = false
-			this._enoTimer = 0
-			this._droppedEno ++
-			this._lostFoul ++
-		}
-		if (this._poly > 0 && this._eno) {
-			this._eno = false
-			this._enoTimer = 0
-			this._droppedEno ++
+	addEvent() {
+		const lastAdded = this._toAdd.length > 0 ? this._toAdd[this._toAdd.length - 1] : null
+
+		if (this.gaugeValuesChanged(lastAdded)) {
+			this._toAdd.push({
+				type: BLM_GAUGE_EVENT,
+				timestamp: this._currentTimestamp,
+				insertAfter: this._normalizeIndex,
+				astralFire: this._astralFireStacks,
+				umbralIce: this._umbralIceStacks,
+				umbralHearts: this._umbralHeartStacks,
+				enochian: this._hasEnochian,
+				polyglot: this._hasPolyglot,
+				lastGaugeEvent: this._lastAdded,
+			})
+			this._lastAdded = this._toAdd[this._toAdd.length - 1]
 		}
 	}
 
-	_onBegin(event) {
-		this.enoTimerCheck(event)
+	constructor(...args) {
+		super(...args)
+		this.addHook('complete', this._onComplete)
 	}
 
-	_onCast(event) {
-		const abilityId = event.ability.guid
-		this.enoTimerCheck(event)
-
-		//check if eno is active and update the eno timer for foul. Check for foul overwriting.
-		if (this._eno) {
-			const enoRunTime = event.timestamp - this._enoTimer
-			const numberOfFouls = Math.floor(enoRunTime/ENOCHIAN_DURATION_REQUIRED)
-			if (numberOfFouls > 0) {
-				const offSet = enoRunTime % ENOCHIAN_DURATION_REQUIRED
-				this._enoTimer = event.timestamp - offSet
-				this._poly ++
-				if (this._poly > 1) {
-					this._poly = 1
-					this._overwrittenFoul ++
-				}
+	normalise(events) {
```

Comment:
I'm heavily questioning the use of normalise for this. Any reason the events can't be sent out with `this.parser.fabricateEvent`?

Reply:
As per discussion on discord, going to let this go for now. Need to re-consider how some GCD based calculations are run so we can pull some of the normalise logic out of normalisation in the future.